Feature/support feature flags

imhmdb opened the pull request:
Follows spec for implementing feature flags support:
https://github.com/PostHog/posthog.com/pull/1455/files

yakkomajuri reviewed (changes requested):
Hey, thanks a lot for this! Appreciate you submitting it upstream. 

Left one inline comment, but overall looking good. Some thoughts:

1. I'm happy to merge this with only `LibCurl` support
2. Some tests that mock the endpoint response and verify that the code handles the response appropriately should be enough. 
3. This implementation appears _correct_ and is an improvement over what we currently have (no feature flags), so it should be valid to merge. However, I'd appreciate if you could consider adding support for `is_simple_flag` (see [Python lib](https://github.com/PostHog/posthog-python/blob/a139795a74429853a3d4aabf512683a188208c56/posthog/client.py#L287)). I'm adding feature flags to `posthog-node` this week and will write a spec for implementing flags in other libraries. Essentially, for `is_simple_flag`, we use a Personal API Key to poll for flags and if they're "simple" (only rollout percentage, no filters), we calculate them locally for speed. This can definitely be a separate PR and I'm happy to help.

yakkomajuri commented:
[Node is done](https://github.com/PostHog/posthog-node/pull/29) and the ["spec" is written](https://github.com/PostHog/posthog.com/pull/1455)

@imhmdb are you planning on making the changes here? totally understand if not, just checking

imhmdb commented:
> [Node is done](https://github.com/PostHog/posthog-node/pull/29) and the ["spec" is written](https://github.com/PostHog/posthog.com/pull/1455)
> 
> @imhmdb are you planning on making the changes here? totally understand if not, just checking

imhmdb commented:
> [Node is done](https://github.com/PostHog/posthog-node/pull/29) and the ["spec" is written](https://github.com/PostHog/posthog.com/pull/1455)
> 
> @imhmdb are you planning on making the changes here? totally understand if not, just checking

of course, I'll give it a shot today and see how it goes.

imhmdb commented:
I think this only needs to have tests fixed with mocks

imhmdb commented:
Done fixing tests with the mocked response of simple flags.

yakkomajuri commented:
Awesome, thanks a lot! I'll give this a review latest by tomorrow

yakkomajuri commented on lib/Client.php:
```diff
@@ -203,4 +305,42 @@ private function message($msg)
+    private function loadFeatureFlags(): void
+    {
+        $response = $this->httpClient->sendRequest(
+            '/api/feature_flag',
```
We've made a change since you last edited this, so this should now be `/api/feature_flag?token=project_api_key_here`.

No way you could have known though!

yakkomajuri commented on lib/Client.php:
```diff
@@ -203,4 +305,42 @@ private function message($msg)
+            $this->featureFlags = [];
+        }
+
+        $this->featureFlags = $responseBody['results'];
```
Another change we made is that this should filter by active flags.

The `active` property is in `$responseBody['results'][i]['active']`

yakkomajuri commented on test/PostHogTest.php:
```diff
@@ -64,6 +73,21 @@ public function testIdentify(): void
+    {
+        $this->assertIsArray(PostHog::fetchEnabledFeatureFlags('user-id'));
+    }
+
```
could we add a test for `isSimpleFlagEnabled`? 

Flag of key `a` with distinct id `b` should be on for rollout percentage `42` but off for rollout percentage `40`

yakkomajuri reviewed (changes requested):
This is awesome! Follows the spec perfectly. I'm about to test it locally and will report back.

Thanks again for doing this.

One final comment that just came to mind: should we also error out if people chose not to use libcurl and they try to use feature flags?

yakkomajuri commented on lib/Client.php:
```diff
@@ -203,4 +305,42 @@ private function message($msg)
+
+    private function loadFeatureFlags(): void
+    {
+        $response = $this->httpClient->sendRequest(
```
Ok so I tried this locally and it didn't work. I _suspect_ this is a potential problem. We're sending a POST request here, when it should actually be a GET request.

yakkomajuri reviewed (changes requested):
Having tried this and re-read the code - there are a few issues. First is a POST request to `api/feature_flag`. Second is that we're missing a poller. We shouldn't load feature flags once, we should load them regularly at an interval in the background.

imhmdb commented:
> This is awesome! Follows the spec ~~perfectly~~ for the most part (see below). I'm about to test it locally and will report back.
> 
> 
> 
> Thanks again for doing this.
> 
> 
> 
> One final comment that just came to mind: should we also error out if people chose not to use libcurl and they try to use feature flags?


Hi, @yakkomajuri 

Regarding the libcurl error, when I reimplemented according to spec I moved the feature flags from libcurl consumer into the client, not sure if you mean ext-curl and not the libcurl-posthog-consumer here, cuz that would be a valid case to error on. 


Regarding the poller, I wasn't sure how can I keep polling at an interval when php doesn't supports threads or pools, have this poller been implemented in a php-like client that I can take insights from?

Regarding the errors on the POST request, this sure is weird As I made sure the tests were working on my production environment keys instead of debug ones, but maybe the tests were also broken 😂


I'll sure take a second look and try to debug and fix the addressed issues.

Side note: I just got out of surgery so it might take me a week or two of recovery until I can get my hands on this thing again, just letting you guys know.

yakkomajuri commented:
I could also be at fault here - not a PHP dev, so if anything I say sounds crazy, it probably is. 

Also, get better! No rush!

I might also pick this up if I get the time. 

Thanks again for all your work @imhmdb

yakkomajuri commented:
Closing in favor of #18

@posthog-bot please add @imhmdb for code

Twixes commented:
@posthog-bot please add @imhmdb for code

Twixes commented:
@posthog-bot please add @imhmdb for code
